robot-agent/packageMonitor.js

```diff
@@ -0,0 +1,168 @@
+  constructor() {
+    super();
+    log.debug('PackageMonitor created');
+    this.installedPackages = {};
+    this.mqttSync = null;
+    this.AGENT_PREFIX = null;
+    this.initialized = false;
+    // for late subscribers raise events for already running packages
+    this.on('newListener', (event, listener) => {
+      log.debug(`New listener for event: ${event}`);
+      if (event === 'packageStarted') {
+        Object.entries(this.installedPackages).forEach(([pkgName, pkgData]) => {
+          if (pkgData.running) {
+            log.info(`Raising packageStarted event for already running package ${pkgName}`);
+            setImmediate(() => listener(pkgName, pkgData.pid));
+          }
+        });
+      }
+      if (event === 'packageStopped') {
+        Object.entries(this.installedPackages).forEach(([pkgName, pkgData]) => {
+          if (pkgData.running === false) {
+            log.info(`Raising packageStopped event for already stopped package ${pkgName}`);
+            setImmediate(() => listener(pkgName));
+          }
+        });
+      }
+    });
+  }
+
+  /**
+   * Initializes the PackageMonitor
+   * Sets up MQTT subscriptions to monitor package status changes and waits for
+   * the initial heartbeat to mark the monitor as initialized.
+   * 
+   * @param {Object} mqttSync 
+   * @param {string} agentPrefix
+   */
+  init(mqttSync, agentPrefix) {
+    this.mqttSync = mqttSync;
+    this.AGENT_PREFIX = agentPrefix;
+    log.info('Initializing PackageMonitor');
+    this.mqttSync.subscribe(`${this.AGENT_PREFIX}/status/runningPackages`);
+    this.mqttSync.data.subscribePathFlat(`${this.AGENT_PREFIX}/status/runningPackages/+scope/+capName/+version`,
+      async (value, topic, matched, tags) => {
+        const {scope, capName, version} = matched;
+        const pkgName = `${scope}/${capName}`;
+
+        if (!value || value === 'false') {
+          log.info(`Package ${pkgName} is no longer running (via MQTT)`);
+          if (this.installedPackages?.[pkgName]?.running) {
+            this.installedPackages[pkgName].running = false;
+            this.installedPackages[pkgName].pid = null;
+            this.emit('packageStopped', pkgName);
+          }
+        } else {
+          log.info(`Package ${pkgName} is running (via MQTT)`);
+          if (!this.installedPackages?.[pkgName]?.running) {
+            const pkgPid = await this.getPackagePid(pkgName);
+            this.installedPackages[pkgName] ||= {};
+            this.installedPackages[pkgName].running = true;
+            this.installedPackages[pkgName].pid = pkgPid;
+            this.emit('packageStarted', pkgName, pkgPid);
+          }
+        }
+      }
+    );
+    this.refreshAllInstalledPackages();
+    this.mqttSync.waitForHeartbeatOnce(() => {
+      log.info('PackageMonitor heartbeat received');
+      this.initialized = true; // Set initialized state
+      log.info('PackageMonitor initialized');
+      this.emit('initialized');
+    });
```

Comment:
This whole `initialized` logic is now obsolete it seems (no longer used anywhere if I see correctly).